### PR TITLE
feat: http/https proxy support

### LINKED ISSUE #840: [REQUEST] Add proxy param for interactions.Client
### Describe the feature.

In some areas, we can't directly connect to Discord server, so we need proxy.

In discord.py, we use the code below to connect proxy server:
```py
bot = commands.Bot(..., proxy="http://127.0.0.1:7890")
```
or
```py
client = discord.Client(
    connector=ProxyConnector(
        proxy_type=ProxyType.SOCKS5,
        host='127.0.0.1',
        port=7890,
        rdns=True
    ),
    ...
)
client.run(BOT_TOKEN)
```

I think in interactions.py, we can use the code below to connect:
```py
interactions.Client(BOT_TOKEN, proxy="http://127.0.0.1:7890")
```

### Code of Conduct

- [X] I agree to follow the contribution requirements.

### PR DESCRIPTION
**Pull request by V3ntus**
## About

Add simple http/https proxy support with authentication.
Still WIP

## Checklist

- [ ] I've ran `pre-commit` to format and lint the change(s) made.
- [ ] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [x] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent): #840 
- [x] I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

File: interactions/client/bot.py [hunk 1]
```diff
@@ -115,6 +120,26 @@ def __init__(
...
+                _re_proxy = re.search(
+                    "^((?P<scheme>[^:/?#]+):(?=//))?(//)?(((?P<login>[^:]+)(?::(?P<password>[^@]+)?)?@)?(?P<host>[^@/?#:]*)(?::(?P<port>\d+)?)?)?",
+                    __proxy,
+                )  # noqa: E501
+                self._proxy = ProxyConfig(
+                    host=_re_proxy["host"],
+                    port=_re_proxy["port"],
+                    user=_re_proxy["login"],
+                    password=_re_proxy["password"],
+                )
```
Comment (AstreaTSS) [inline interactions/client/bot.py]: Is there anything telling this to error out if the regex fails to match anything?

Comment (sillyElvy) [reply interactions/client/bot.py]: Also the flake8 fails on the regex

Comment (V3ntus) [reply interactions/client/bot.py]: Line too long
Flake8 might be lying, it was validated. But idk, I haven't tested

File: interactions/api/gateway/client.py [hunk 2]
```diff
@@ -111,6 +150,7 @@ def __init__(
             "timeout": 60,
             "autoclose": False,
             "compress": 0,
+            "proxy": _proxy if isinstance(_proxy, ProxyConfig) else None,
```
Comment (FayeDel) [inline interactions/api/gateway/client.py]: FYI (or a tip/suggestion/advice)
the proxy arg here have to either be string or None, not proxyconfig objects
(+, if you have an BasicAuth object, you can do `"proxy_auth": <your auth object>` if it exists)

Comment (V3ntus) [reply interactions/api/gateway/client.py]: (yeah I noticed that after I pushed)
I'll correct a lot later when I'm done with work

File: interactions/api/gateway/client.py [hunk 3]
```diff
@@ -111,7 +113,10 @@ def __init__(
             "timeout": 60,
             "autoclose": False,
             "compress": 0,
+            "proxy": str(proxy) if isinstance(proxy, ProxyConfig) else None,
+            "proxy_auth": proxy.auth if isinstance(proxy, ProxyConfig) else None,
         }
+        print(self._options)
```
Comment (sillyElvy) [inline interactions/api/gateway/client.py]: No

**Review by sillyElvy — CHANGES_REQUESTED** [review]

File: interactions/client/bot.py [hunk 4]
```diff
@@ -89,11 +91,26 @@ def __init__(
...
+        # proxy? : Optional[Union[ProxyConfig, str]]
+        #     Sets the proxy configuration of the client.
+
+        if __proxy := kwargs.get("proxy"):
+            if isinstance(__proxy, str):
+                self._proxy = ProxyConfig(__proxy)
+            elif isinstance(__proxy, ProxyConfig):
+                self._proxy = __proxy
+            else:
+                log.error(
```
Comment (sillyElvy) [inline interactions/client/bot.py]: Raise a actual error suits better I think

File: interactions/api/gateway/client.py (hunk 3 above)
Comment (V3ntus) [reply interactions/api/gateway/client.py]: ah yes i should've searched for my debug print statements, huh

File: interactions/client/bot.py [hunk 5]
```diff
@@ -89,11 +91,26 @@ def __init__(
...
+        # proxy? : Optional[Union[ProxyConfig, str]]
+        #     Sets the proxy configuration of the client.
+
+        if __proxy := kwargs.get("proxy"):
+            if isinstance(__proxy, str):
+                self._proxy = ProxyConfig(__proxy)
+            elif isinstance(__proxy, ProxyConfig):
+                self._proxy = __proxy
+            else:
+                raise Exception(
```
Comment (sillyElvy) [inline interactions/client/bot.py:103]: Why not `LibraryException`?

**Pull request closed without merging**